op-txproxy/conditional_txs.go

```diff
@@ -0,0 +1,143 @@
+package op_txproxy
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/metrics"
+	"github.com/ethereum-optimism/optimism/op-service/predeploys"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/time/rate"
+)
+
+var (
+	// errs
+	rateLimitErr             = &rpc.JsonError{Message: "rate limited", Code: params.TransactionConditionalRejectedErrCode}
+	endpointDisabledErr      = &rpc.JsonError{Message: "endpoint disabled", Code: params.TransactionConditionalRejectedErrCode}
+	missingAuthenticationErr = &rpc.JsonError{Message: "missing authentication", Code: params.TransactionConditionalRejectedErrCode}
+	entrypointSupportErr     = &rpc.JsonError{Message: "only 4337 Entrypoint contract support", Code: params.TransactionConditionalRejectedErrCode}
+)
+
+type ConditionalTxService struct {
+	log log.Logger
+	cfg *CLIConfig
+
+	limiter             *rate.Limiter
+	backend             client.RPC
+	entrypointAddresses map[common.Address]bool
+
+	costSummary prometheus.Summary
+	requests    prometheus.Counter
+	failures    *prometheus.CounterVec
+}
+
+func NewConditionalTxService(ctx context.Context, log log.Logger, m metrics.Factory, cfg *CLIConfig) (*ConditionalTxService, error) {
+	rpc, err := client.NewRPC(ctx, log, cfg.SendRawTransactionConditionalBackend)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial backend %s: %w", cfg.SendRawTransactionConditionalBackend, err)
+	}
+
+	rpcMetrics := metrics.MakeRPCClientMetrics("backend", m)
+	backend := client.NewInstrumentedRPC(rpc, &rpcMetrics)
+
+	limiter := rate.NewLimiter(params.TransactionConditionalMaxCost, int(cfg.SendRawTransactionConditionalRateLimit))
```

Comment:
Need to update this based on this discussion: https://github.com/ethereum-optimism/op-geth/pull/330#discussion_r1752214483